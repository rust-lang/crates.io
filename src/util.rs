pub use self::io_util::{read_fill, read_le_u32};
pub use self::request_helpers::*;
pub use crates_io_database::utils::token;

pub mod diesel;
pub mod errors;
mod io_util;
mod request_helpers;
pub mod string_excl_null;
pub mod tracing;
