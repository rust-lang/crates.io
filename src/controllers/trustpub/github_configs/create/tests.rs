use crate::tests::builders::CrateBuilder;
use crate::tests::util::{RequestHelper, Response, TestApp};
use anyhow::anyhow;
use bytes::Bytes;
use crates_io_database::models::token::{CrateScope, EndpointScope};
use crates_io_database::schema::{emails, trustpub_configs_github};
use crates_io_github::{GitHubError, GitHubUser, MockGitHubClient};
use diesel::prelude::*;
use diesel_async::RunQueryDsl;
use insta::{assert_json_snapshot, assert_snapshot};
use serde_json::json;

const URL: &str = "/api/v1/trusted_publishing/github_configs";

const CRATE_NAME: &str = "foo";

fn simple_github_mock() -> MockGitHubClient {
    let mut github_mock = MockGitHubClient::new();
    github_mock.expect_get_user().returning(|login, _| {
        Ok(GitHubUser {
            avatar_url: None,
            email: None,
            id: 42,
            login: login.into(),
            name: None,
        })
    });
    github_mock
}

async fn run_test(payload: impl Into<Bytes>) -> (TestApp, Response<()>) {
    async fn inner(payload: Bytes) -> (TestApp, Response<()>) {
        let (app, _client, cookie_client) = TestApp::full()
            .with_github(simple_github_mock())
            .with_user()
            .await;

        let mut conn = app.db_conn().await;

        CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
            .build(&mut conn)
            .await
            .unwrap();

        (app, cookie_client.post::<()>(URL, payload).await)
    }

    inner(payload.into()).await
}

#[tokio::test(flavor = "multi_thread")]
async fn test_happy_path() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let (app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    assert_snapshot!(app.emails_snapshot().await);

    let mut conn = app.db_conn().await;
    let config_ids = trustpub_configs_github::table
        .select(trustpub_configs_github::id)
        .get_results::<i32>(&mut conn)
        .await?;

    assert_eq!(config_ids.len(), 1);
    assert_eq!(config_ids[0], 1);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_happy_path_with_environment() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": "production",
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_empty_body() -> anyhow::Result<()> {
    let (_app, response) = run_test("").await;
    assert_snapshot!(response.status(), @"415 Unsupported Media Type");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Expected request with `Content-Type: application/json`"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_empty_json_object() -> anyhow::Result<()> {
    let (_app, response) = run_test("{}").await;
    assert_snapshot!(response.status(), @"422 Unprocessable Entity");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Failed to deserialize the JSON body into the target type: missing field `github_config` at line 1 column 2"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_owner() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "ยง$%&",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Invalid GitHub repository owner name"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_repo() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "@foo",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Invalid GitHub repository name"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_workflow() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "workflows/ci.json",
            "environment": null,
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Workflow filename must end with `.yml` or `.yaml`"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_environment() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": "",
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Environment name may not be empty (use `null` to omit)"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_unauthenticated() -> anyhow::Result<()> {
    let (app, client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this action requires authentication"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_legacy_token_auth() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_token()
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_with_trusted_publishing_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from(CRATE_NAME).unwrap()]),
            Some(vec![EndpointScope::TrustedPublishing]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_without_trusted_publishing_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from(CRATE_NAME).unwrap()]),
            Some(vec![EndpointScope::PublishUpdate]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this token does not have the required permissions to perform this action"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_with_wrong_crate_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from("other-crate").unwrap()]),
            Some(vec![EndpointScope::TrustedPublishing]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this token does not have the required permissions to perform this action"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_with_wildcard_crate_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from("*").unwrap()]),
            Some(vec![EndpointScope::TrustedPublishing]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_missing_crate() -> anyhow::Result<()> {
    let (_app, _client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"404 Not Found");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"crate `foo` does not exist"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_non_owner() -> anyhow::Result<()> {
    let (app, _client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let other_client = app.db_new_user("other_user").await;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = other_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"You are not an owner of this crate"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_unknown_github_user() -> anyhow::Result<()> {
    let mut github_mock = MockGitHubClient::new();
    github_mock
        .expect_get_user()
        .returning(|login, _| Err(GitHubError::NotFound(anyhow!("User {} not found", login))));

    let (app, _client, cookie_client) = TestApp::full().with_github(github_mock).with_user().await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Unknown GitHub user or organization"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_github_error() -> anyhow::Result<()> {
    let mut github_mock = MockGitHubClient::new();
    github_mock
        .expect_get_user()
        .returning(|_, _| Err(GitHubError::Other(anyhow!("Internal Server Error"))));

    let (app, _client, cookie_client) = TestApp::full().with_github(github_mock).with_user().await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"500 Internal Server Error");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Internal Server Error"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_unverified_email() -> anyhow::Result<()> {
    let (app, _client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let mut conn = app.db_conn().await;

    diesel::update(emails::table.filter(emails::user_id.eq(cookie_client.as_model().id)))
        .set(emails::verified.eq(false))
        .execute(&mut conn)
        .await?;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"You must verify your email address to create a Trusted Publishing config"}]}"#);

    Ok(())
}
