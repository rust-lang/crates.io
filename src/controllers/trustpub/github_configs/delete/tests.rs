use crate::tests::builders::CrateBuilder;
use crate::tests::util::{RequestHelper, TestApp};
use crates_io_database::models::Crate;
use crates_io_database::models::trustpub::{GitHubConfig, NewGitHubConfig};
use crates_io_database::schema::trustpub_configs_github;
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};
use http::StatusCode;
use insta::assert_snapshot;
use serde_json::json;

const BASE_URL: &str = "/api/v1/trusted_publishing/github_configs";
const CRATE_NAME: &str = "foo";

fn delete_url(id: i32) -> String {
    format!("{BASE_URL}/{id}")
}

async fn create_crate(conn: &mut AsyncPgConnection, author_id: i32) -> anyhow::Result<Crate> {
    CrateBuilder::new(CRATE_NAME, author_id).build(conn).await
}

async fn create_config(conn: &mut AsyncPgConnection, crate_id: i32) -> QueryResult<GitHubConfig> {
    let config = NewGitHubConfig {
        crate_id,
        repository_owner: "rust-lang",
        repository_owner_id: 42,
        repository_name: "foo-rs",
        workflow_filename: "publish.yml",
        environment: None,
    };

    config.insert(conn).await
}

async fn get_all_configs(conn: &mut AsyncPgConnection) -> QueryResult<Vec<GitHubConfig>> {
    trustpub_configs_github::table
        .select(GitHubConfig::as_select())
        .load::<GitHubConfig>(conn)
        .await
}

/// Delete the config with a valid user that is an owner of the crate.
#[tokio::test(flavor = "multi_thread")]
async fn test_happy_path() -> anyhow::Result<()> {
    let (app, _client, cookie_client) = TestApp::full().with_user().await;
    let mut conn = app.db_conn().await;

    let krate = create_crate(&mut conn, cookie_client.as_model().id).await?;
    let config = create_config(&mut conn, krate.id).await?;

    let response = cookie_client.delete::<()>(&delete_url(config.id)).await;
    assert_eq!(response.status(), StatusCode::NO_CONTENT);
    assert_eq!(response.text(), "");

    // Verify the config was deleted from the database
    let configs = get_all_configs(&mut conn).await?;
    assert_eq!(configs.len(), 0);

    // Verify emails were sent to crate owners
    assert_snapshot!(app.emails_snapshot().await);

    Ok(())
}

/// Try to delete the config with an unauthenticated client.
#[tokio::test(flavor = "multi_thread")]
async fn test_unauthenticated() -> anyhow::Result<()> {
    let (app, client, cookie_client) = TestApp::full().with_user().await;
    let mut conn = app.db_conn().await;

    let krate = create_crate(&mut conn, cookie_client.as_model().id).await?;
    let config = create_config(&mut conn, krate.id).await?;

    let response = client.delete::<()>(&delete_url(config.id)).await;
    assert_eq!(response.status(), StatusCode::FORBIDDEN);
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this action requires authentication"}]}"#);

    // Verify the config was not deleted
    let configs = get_all_configs(&mut conn).await?;
    assert_eq!(configs.len(), 1);

    // Verify no emails were sent to crate owners
    assert_eq!(app.emails().await.len(), 0);

    Ok(())
}

/// Try to delete the config with API token authentication.
#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full().with_token().await;
    let mut conn = app.db_conn().await;

    let krate = create_crate(&mut conn, cookie_client.as_model().id).await?;
    let config = create_config(&mut conn, krate.id).await?;

    let response = token_client.delete::<()>(&delete_url(config.id)).await;
    assert_eq!(response.status(), StatusCode::FORBIDDEN);
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this action can only be performed on the crates.io website"}]}"#);

    // Verify the config was not deleted
    let configs = get_all_configs(&mut conn).await?;
    assert_eq!(configs.len(), 1);

    // Verify no emails were sent to crate owners
    assert_eq!(app.emails().await.len(), 0);

    Ok(())
}

/// Try to delete a config that does not exist.
#[tokio::test(flavor = "multi_thread")]
async fn test_config_not_found() -> anyhow::Result<()> {
    let (app, _client, cookie_client) = TestApp::full().with_user().await;

    let response = cookie_client.delete::<()>(&delete_url(42)).await;
    assert_eq!(response.status(), StatusCode::NOT_FOUND);
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Not Found"}]}"#);

    // Verify no emails were sent to crate owners
    assert_eq!(app.emails().await.len(), 0);

    Ok(())
}

/// Try to delete the config with a user who is not an owner of the crate.
#[tokio::test(flavor = "multi_thread")]
async fn test_non_owner() -> anyhow::Result<()> {
    let (app, _client, cookie_client) = TestApp::full().with_user().await;
    let mut conn = app.db_conn().await;

    let krate = create_crate(&mut conn, cookie_client.as_model().id).await?;
    let config = create_config(&mut conn, krate.id).await?;

    // Create another user who is not an owner of the crate
    let other_client = app.db_new_user("other_user").await;

    let response = other_client.delete::<()>(&delete_url(config.id)).await;
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"You are not an owner of this crate"}]}"#);

    // Verify the config was not deleted
    let configs = get_all_configs(&mut conn).await?;
    assert_eq!(configs.len(), 1);

    // Verify no emails were sent to crate owners
    assert_eq!(app.emails().await.len(), 0);

    Ok(())
}

/// Try to delete the config with a user that is part of a team that owns
/// the crate.
#[tokio::test(flavor = "multi_thread")]
async fn test_team_owner() -> anyhow::Result<()> {
    let (app, _client) = TestApp::full().empty().await;
    let mut conn = app.db_conn().await;

    let user = app.db_new_user("user-org-owner").await;
    let user2 = app.db_new_user("user-one-team").await;

    let krate = create_crate(&mut conn, user.as_model().id).await?;
    let config = create_config(&mut conn, krate.id).await?;

    let body = json!({ "owners": ["github:test-org:all"] }).to_string();
    let response = user.put::<()>("/api/v1/crates/foo/owners", body).await;
    assert_eq!(response.status(), StatusCode::OK);

    let response = user2.delete::<()>(&delete_url(config.id)).await;
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"You are not an owner of this crate"}]}"#);

    // Verify the config was not deleted
    let configs = get_all_configs(&mut conn).await?;
    assert_eq!(configs.len(), 1);

    // Verify no emails were sent to crate owners
    assert_eq!(app.emails().await.len(), 0);

    Ok(())
}
