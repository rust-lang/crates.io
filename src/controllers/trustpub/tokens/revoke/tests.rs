use crate::tests::util::{MockTokenUser, RequestHelper, TestApp};
use chrono::{TimeDelta, Utc};
use crates_io_database::models::trustpub::NewToken;
use crates_io_database::schema::trustpub_tokens;
use crates_io_trustpub::access_token::AccessToken;
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};
use insta::assert_compact_debug_snapshot;
use insta::assert_snapshot;
use secrecy::ExposeSecret;
use sha2::Sha256;
use sha2::digest::Output;

const URL: &str = "/api/v1/trusted_publishing/tokens";

fn generate_token() -> (String, Output<Sha256>) {
    let token = AccessToken::generate();
    (token.finalize().expose_secret().to_string(), token.sha256())
}

async fn new_token(conn: &mut AsyncPgConnection, crate_id: i32) -> QueryResult<String> {
    let (token, hashed_token) = generate_token();

    let new_token = NewToken {
        expires_at: Utc::now() + TimeDelta::minutes(30),
        hashed_token: hashed_token.as_slice(),
        crate_ids: &[crate_id],
    };

    new_token.insert(conn).await?;

    Ok(token)
}

async fn all_crate_ids(conn: &mut AsyncPgConnection) -> QueryResult<Vec<Vec<Option<i32>>>> {
    trustpub_tokens::table
        .select(trustpub_tokens::crate_ids)
        .load(conn)
        .await
}

#[tokio::test(flavor = "multi_thread")]
async fn test_happy_path() -> anyhow::Result<()> {
    let (app, _client) = TestApp::full().empty().await;
    let mut conn = app.db_conn().await;

    let token1 = new_token(&mut conn, 1).await?;
    let _token2 = new_token(&mut conn, 2).await?;
    assert_compact_debug_snapshot!(all_crate_ids(&mut conn).await?, @"[[Some(1)], [Some(2)]]");

    let header = format!("Bearer {}", token1);
    let token_client = MockTokenUser::with_auth_header(header, app.clone());

    let response = token_client.delete::<()>(URL).await;
    assert_snapshot!(response.status(), @"204 No Content");
    assert_eq!(response.text(), "");

    // Check that the token is deleted
    assert_compact_debug_snapshot!(all_crate_ids(&mut conn).await?, @"[[Some(2)]]");

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_missing_authorization_header() -> anyhow::Result<()> {
    let (_app, client) = TestApp::full().empty().await;

    let response = client.delete::<()>(URL).await;
    assert_snapshot!(response.status(), @"401 Unauthorized");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Missing authorization header"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_authorization_header_format() -> anyhow::Result<()> {
    let (app, _client) = TestApp::full().empty().await;

    // Create a client with an invalid authorization header (missing "Bearer " prefix)
    let header = "invalid-format".to_string();
    let token_client = MockTokenUser::with_auth_header(header, app.clone());

    let response = token_client.delete::<()>(URL).await;
    assert_snapshot!(response.status(), @"401 Unauthorized");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Invalid authorization header"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_token_format() -> anyhow::Result<()> {
    let (app, _client) = TestApp::full().empty().await;

    // Create a client with an invalid token format
    let header = "Bearer invalid-token".to_string();
    let token_client = MockTokenUser::with_auth_header(header, app.clone());

    let response = token_client.delete::<()>(URL).await;
    assert_snapshot!(response.status(), @"401 Unauthorized");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Invalid authorization header"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_non_existent_token() -> anyhow::Result<()> {
    let (app, _client) = TestApp::full().empty().await;

    // Generate a valid token format, but it doesn't exist in the database
    let (token, _) = generate_token();
    let header = format!("Bearer {}", token);
    let token_client = MockTokenUser::with_auth_header(header, app.clone());

    // The request should succeed with 204 No Content even though the token doesn't exist
    let response = token_client.delete::<()>(URL).await;
    assert_snapshot!(response.status(), @"204 No Content");
    assert_eq!(response.text(), "");

    Ok(())
}
