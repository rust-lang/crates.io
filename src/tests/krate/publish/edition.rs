use crate::tests::builders::PublishBuilder;
use crate::tests::util::insta::{any_id_redaction, id_redaction};
use crate::tests::util::{RequestHelper, TestApp};
use insta::{assert_json_snapshot, assert_snapshot};

#[tokio::test(flavor = "multi_thread")]
async fn test_edition_is_saved() {
    let (_app, _, _, token) = TestApp::full().with_token().await;

    let manifest = r#"
        [package]
        name = "foo"
        version = "1.0.0"
        description = "description"
        license = "MIT"
        edition = "2021"
        rust-version = "1.0"
    "#;
    let pb = PublishBuilder::new("foo", "1.0.0").custom_manifest(manifest);
    let response = token.publish_crate(pb).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), {
        ".crate.created_at" => "[datetime]",
        ".crate.updated_at" => "[datetime]",
    });

    let response = token.get::<()>("/api/v1/crates/foo/1.0.0").await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), {
        ".version.id" => any_id_redaction(),
        ".version.created_at" => "[datetime]",
        ".version.updated_at" => "[datetime]",
        ".version.published_by.id" => id_redaction(token.as_model().user_id),
        ".version.audit_actions[].time" => "[datetime]",
        ".version.audit_actions[].user.id" => id_redaction(token.as_model().user_id),
    });
}
