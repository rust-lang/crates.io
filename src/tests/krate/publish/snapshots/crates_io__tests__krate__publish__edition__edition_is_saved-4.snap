---
source: src/tests/krate/publish/edition.rs
expression: response.json()
---
{
  "version": {
    "audit_actions": [
      {
        "action": "publish",
        "time": "[datetime]",
        "user": {
          "avatar": null,
          "id": "[id]",
          "login": "foo",
          "name": null,
          "url": "https://github.com/foo"
        }
      }
    ],
    "bin_names": [],
    "checksum": "7008c5c6b6cbf418070a998fad244bbde3ca712148e61e4d4b0d257682e30ad2",
    "crate": "foo",
    "crate_size": 172,
    "created_at": "[datetime]",
    "description": "description",
    "dl_path": "/api/v1/crates/foo/1.0.0/download",
    "documentation": null,
    "downloads": 0,
    "edition": "2021",
    "features": {},
    "has_lib": false,
    "homepage": null,
    "id": "[id]",
    "lib_links": null,
    "license": "MIT",
    "links": {
      "authors": "/api/v1/crates/foo/1.0.0/authors",
      "dependencies": "/api/v1/crates/foo/1.0.0/dependencies",
      "version_downloads": "/api/v1/crates/foo/1.0.0/downloads"
    },
    "num": "1.0.0",
    "published_by": {
      "avatar": null,
      "id": "[id]",
      "login": "foo",
      "name": null,
      "url": "https://github.com/foo"
    },
    "readme_path": "/api/v1/crates/foo/1.0.0/readme",
    "repository": null,
    "rust_version": "1.0",
    "updated_at": "[datetime]",
    "yank_message": null,
    "yanked": false
  }
}
